File: interpreter/src/globals.py
from typing import TYPE_CHECKING
from typing import Self

from src.runtime.callable import Callable
from src.runtime.values import MK_NUMBER
from src.runtime.values import AlgebraicType
from src.runtime.values import AlgebraicValue


if TYPE_CHECKING:
    from src.interpreter import Interpreter


class TailleFunction(Callable):
    def arity(self: Self):
        return 1

    def call(
        self: Self, _: "Interpreter", args: list[AlgebraicValue]
    ) -> AlgebraicValue:
        data = args[0]
        if data.type == AlgebraicType.STRING:
            data_value: str = data.value  # type: ignore
            return MK_NUMBER(len(data_value))
        if data.type == AlgebraicType.ARRAY:
            data_value: list[AlgebraicValue] = data.value  # type: ignore
            return MK_NUMBER(len(data_value))
        return MK_NUMBER(0)

    def __str__(self: Self):
        return "<fonction @taille>"

File: qlang-rs/src/globals.rs
use std::rc::Rc;

use crate::callable::{reset_anon_counter, Callable};
use crate::environment::Environment;
use crate::error::QError;
use crate::interpreter::Interpreter;
use crate::values::Value;

struct TailleFunction;

impl Callable for TailleFunction {
    fn arity(&self) -> usize {
        1
    }

    fn call(&self, _interpreter: &mut Interpreter, mut args: Vec<Value>) -> Result<Value, QError> {
        match args.remove(0) {
            Value::Array(items) => Ok(Value::Number(items.borrow().len() as f64)),
            other => Err(QError::runtime(format!(
                "'taille' attend un tableau, reçu '{}'",
                other.type_name()
            ))),
        }
    }

    fn name(&self) -> String {
        "taille".to_string()
    }
}

/// Builds a fresh global environment pre-populated with the standard
/// library available to every qlang program (currently just `taille`).
pub fn make_global_env() -> Environment {
    reset_anon_counter();

    let env = Environment::new(None);
    env.declare_variable("taille", Value::Function(Rc::new(TailleFunction)))
        .expect("declaring a builtin in a fresh environment cannot fail");

    env
}
